### PR TITLE
Added support for saving tiff files > 4GB :)

### PR DESCRIPTION
**Pull request by sammyhansali**
# References and relevant issues

#7195 

# Description

Add support for bigtiff by catching the relevant error raised by tifffile if the saved data will be larger than 4GB, then attempting the write again with the bigtiff flag set to true. (The 4GB limit is for content *after* compression, so we can't just check the size of the data.)

This PR also adds a test for this code path.

### CODE REVIEW

**Comment by psobolewskiPhD** [comment]
Awesome! Thanks for picking this up!
I think a test would be good, but I'm a bit wary of a 4Gb array and file.
Maybe @Czaki has a clever idea to test this?

**Comment by Czaki** [comment]
I do not have a strong opinion. 
The hosted runners that we have used have only 14GB of Storage. https://docs.github.com/en/actions/using-github-hosted-runners/using-github-hosted-runners/about-github-hosted-runners#standard-github-hosted-runners-for-public-repositories

So maybe assume that code works based on tiff specification?

**Review by jni — APPROVED** [review]

**Comment by jni** [comment]
We could add a test that is skipped in CI. Suggested test:

```python
@pytest.mark.skipif(os.getenv('CI', '0') != '0')
@pytest.mark.parametrize('size', [2**21, 2**30, 2**33])
def test_tiff_roundtrip(tmp_path, size):
    idxs = np.random.randint(0, size, size=200)
    data = np.zeros((int(math.cbrt(size)),) * 3, dtype=np.uint8)
    data.ravel()[idxs] = 1
    imsave_tiff(tmp_path / 'temp.tiff', data)
    loaded = tifffile.imread(tmp_path / 'temp.tiff')
    assert set(idxs) == set(np.flatnonzero(loaded))
```

HOWEVER, on my machine, this test just passes??? So I don't know whether it's testing what I think it's testing... 😂 Any thoughts? Is the bigtiff limitation only a thing on Windows? 🤔

**Comment by jni** [comment]
(btw welcome @sammyhansali!!! 😊)

**Comment by jni** [comment]
> it is not a problem of ram, but of space on SSD.

it occurred to me as I was writing things that since we turn on compression, a sparse volume like the one in the test should not take up anywhere near 4GB on disk. Or is that actually the issue with the test passing currently — is the size of the array the limiting factor or the size of the compressed file?

**Comment by Czaki** [comment]
> it occurred to me as I was writing things that since we turn on compression, a sparse volume like the one in the test should not take up anywhere near 4GB on disk. Or is that actually the issue with the test passing currently — is the size of the array the limiting factor or the size of the compressed file?

I forget about compression. But I also do not know how to test if saved file is bigtiff file. And the requirement to use 8GB is also not the best one. 

Test not executed on CI do not save us from errors.

**Comment by jni** [comment]
> Test not executed on CI do not save us from errors.

They at least help with bisecting, for example.

**Comment by jni** [comment]
> is the size of the array the limiting factor or the size of the compressed file?

ah, indeed it must be the size of the compressed file. When I use the example from #7195, which is not compressible, I get a failure:

```
In [9]: data = np.random.randint(low=0, high=2**16, size=(128,4096,4096), dtype='uint16')

In [10]: data.nbytes / 1e9
Out[10]: 4.294967296

In [11]: from napari.utils.io import imsave_tiff

In [12]: imsave_tiff('rand.tiff', data)
---------------------------------------------------------------------------
error                                     Traceback (most recent call last)
<ipython-input-12-e6032e0b0c4e> in ?()
----> 1 imsave_tiff('rand.tiff', data)

~/projects/napari/napari/utils/io.py in ?(filename, data)
    102     else:
    103         # 'compression' kwarg since 2021.6.6; we depend on more recent versions
    104         # now. See:
    105         # https://forum.image.sc/t/problem-saving-generated-labels-in-cellpose-napari/54892/8
--> 106         tifffile.imwrite(filename, data, compression=('zlib', 1))

~/micromamba/envs/all/lib/python3.11/site-packages/tifffile/tifffile.py in ?(file, data, mode, bigtiff, byteorder, imagej, ome, shaped, append, shape, dtype, photometric, planarconfig, extrasamples, volumetric, tile, rowsperstrip, bitspersample, compression, compressionargs, predictor, subsampling, jpegtables, iccprofile, colormap, description, datetime, resolution, resolutionunit, subfiletype, software, metadata, extratags, contiguous, truncate, align, maxworkers, buffersize, returnoffset)
   1448         imagej=imagej,
   1449         ome=ome,
   1450         shaped=shaped,
   1451     ) as tif:
-> 1452         result = tif.write(
   1453             data,
   1454             shape=shape,
   1455             dtype=dtype,

~/micromamba/envs/all/lib/python3.11/site-packages/tifffile/tifffile.py in ?(***failed resolving arguments***)
   3619                 ifd.write(pack(offsetformat, ifdpos + pos))
   3620                 ifd.seek(pos)
   3621                 offset = dataoffset
   3622                 for size in databytecounts:
-> 3623                     ifd.write(pack(offsetformat, offset if size > 0 else 0))
   3624                     offset += size
   3625             else:
   3626                 ifd.write(pack(offsetformat, dataoffset))

~/micromamba/envs/all/lib/python3.11/site-packages/tifffile/tifffile.py in ?(self, fmt, *val)
   4068     def _pack(self, fmt: str, *val: Any) -> bytes:
   4069         """Return values packed to bytes according to format."""
   4070         if fmt[0] not in '<>':
   4071             fmt = self.tiff.byteorder + fmt
-> 4072         return struct.pack(fmt, *val)

error: 'I' format requires 0 <= number <= 4294967295
```

**Comment by jni** [comment]
So, do we merge this without a test?

**Comment by Czaki** [comment]
> HOWEVER, on my machine, this test just passes??? So I don't know whether it's testing what I think it's testing... 😂 Any thoughts? Is the bigtiff limitation only a thing on Windows? 🤔

```quote
    A BigTIFF file is written if the data size is larger than 4 GB less
    32 MB for metadata, and `bigtiff` is not *False*, and `imagej`,
    `truncate` and `compression` are not enabled.
    Unless `byteorder` is specified, the TIFF file byte order is determined
    from the dtype of `data` or the `dtype` argument.
```

After inspecting the code and tiff specification, it is a problem of final tiff size. 


For testing we may use 
```python
with tifffile.TiffFile(image_path) as tiff;
    assert tiff.is_bigtiff
```
That will not lead to load data again in memory. 

Maybe we could improve code to 

```python 
try:
	tifffile.imwrite(filename, data, compression=('zlib', 1))
except: #put here proper exception
	tifffile.imwrite(filename, data, compression=('zlib', 1), bigtiff=True)
```

**Comment by Czaki** [comment]
> @Czaki so do you suggest saving a 4GB tiff to a temporary file and then checking as you suggest?

No. I do not see any reason to use temporary file.

**Comment by jni** [comment]
So what did you mean with:

> For testing we may use
> 
>     with tifffile.TiffFile(image_path) as tiff;
>         assert tiff.is_bigtiff

**Comment by Czaki** [comment]
> So what did you mean with:

I understand that you have asked about normal workflow, not test one. 

now we could write such test. 

```python

def test_large_file(monkeypatch, tmp_path):
    old_write = tifffile.imwrite
    
    def raise_no_bigtiff(*args, **kwargs):
        if "bigtiff" not in kwargs:
            raise struct.error
        old_write(*args, **kwargs)

    data = np.array... #some data generation)

    image_path = str(tmp_path / "data.tif")
    imsave_tiff(image_path, data)

    with tifffile.TiffFile(image_path) as tiff;
        assert tiff.is_bigtiff
```

**Comment by psobolewskiPhD** [comment]
Hi @sammyhansali 
Do you have time to finish this off based on @jni and @Czaki comments?
Or do you rather we get it over the line to merge?
Thanks!
❤️

**Comment by sammyhansali** [comment]
Hi @psobolewskiPhD, I will take care of it! I'll have time to work on it early next week.
I did not see the follow-up requests, sorry about that.

**Comment by psobolewskiPhD** [comment]
No worries! Sounds great! Thanks!😊

**Comment by jni** [comment]
Looks like you [made a small typo](https://github.com/napari/napari/actions/runs/10933300488/job/30351554527?pr=7242#step:13:523). 😊 

<details>

<summary>

```
  E       TypeError: imwrite() got an unexpected keyword argument 'compresion'
```

</summary>

```
  ____________________________ test_imsave_large_file ____________________________
  
  args = ('/home/runner/work/napari/napari/.pytest_tmp/test_imsave_large_file0/data.tif', array([[[15997,  9794, 65180, ..., 38...123,  5936, 21780, ..., 46559, 49238, 14837],
          [ 3682, 37911, 39828, ..., 49052, 33894, 48636]]], dtype=uint16))
  kwargs = {'compression': ('zlib', 1)}
  
      def raise_no_bigtiff(*args, **kwargs):
          if 'bigtiff' not in kwargs:
  >           raise struct.error
  E           struct.error
  
  napari/utils/_tests/test_io.py:80: error
  
  During handling of the above exception, another exception occurred:
  
  monkeypatch = <_pytest.monkeypatch.MonkeyPatch object at 0x7fba85429be0>
  tmp_path = PosixPath('/home/runner/work/napari/napari/.pytest_tmp/test_imsave_large_file0')
  
      def test_imsave_large_file(monkeypatch, tmp_path):
          old_write = tifffile.imwrite
      
          def raise_no_bigtiff(*args, **kwargs):
              if 'bigtiff' not in kwargs:
                  raise struct.error
              old_write(*args, **kwargs)
      
          monkeypatch.setattr(tifffile, 'imwrite', raise_no_bigtiff)
      
          data = np.random.randint(
              low=0, high=2**16, size=(128, 4096, 4096), dtype='uint16'
          )  # 4GB size
      
          # create image and assert image file creation
          image_path = str(tmp_path / 'data.tif')
  >       imsave(image_path, data)
  
  napari/utils/_tests/test_io.py:91: 
  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
  napari/utils/io.py:52: in imsave
      imsave_tiff(filename, data)
  napari/utils/io.py:110: in imsave_tiff
      tifffile.imwrite(
  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
  
  args = ('/home/runner/work/napari/napari/.pytest_tmp/test_imsave_large_file0/data.tif', array([[[15997,  9794, 65180, ..., 38...123,  5936, 21780, ..., 46559, 49238, 14837],
          [ 3682, 37911, 39828, ..., 49052, 33894, 48636]]], dtype=uint16))
  kwargs = {'bigtiff': True, 'compresion': ('zlib', 1)}
  
      def raise_no_bigtiff(*args, **kwargs):
          if 'bigtiff' not in kwargs:
              raise struct.error
  >       old_write(*args, **kwargs)
  E       TypeError: imwrite() got an unexpected keyword argument 'compresion'
  
  napari/utils/_tests/test_io.py:81: TypeError
```

</details>

**Comment by psobolewskiPhD** [comment]
@jni how about making a suggestion for that typo fix, so we can get this merged?

File: napari/utils/io.py [hunk 1]
```diff
@@ -103,7 +104,15 @@ def imsave_tiff(filename, data):
...
         # now. See:
         # https://forum.image.sc/t/problem-saving-generated-labels-in-cellpose-napari/54892/8
-        tifffile.imwrite(filename, data, compression=('zlib', 1))
+        try:
+            tifffile.imwrite(filename, data, compression=('zlib', 1))
+        except struct.error:  # compressed data >4GB
+            tifffile.imwrite(
+                filename,
+                data,
+                compresion=('zlib', 1),
```
Comment (Czaki) [inline napari/utils/io.py]:
```suggestion
                compression=('zlib', 1),
```

**Review by Czaki — APPROVED** [review]

**Pull request merged**